tracing/src/test/java/com/palantir/tracing/TracersTest.java

```diff
@@ -450,6 +457,292 @@ public void testWrapRunnableWithAlternateTraceId_traceStateRestoredToCleared() {
         assertThat(Tracer.hasTraceId()).isFalse();
     }
 
+    @Test
+    public void testWrappingFutureCallback_futureCallbackTraceIsIsolated() throws Exception {
+        FutureCallback<Void> futureCallback = new FutureCallback<Void>() {
+            @Override
+            public void onSuccess(@NullableDecl Void result) {
+                Tracer.startSpan("inside"); // never completed
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                Tracer.startSpan("inside"); // never completed
+            }
+        };
+
+        ListeningExecutorService listeningExecutorService =
+                MoreExecutors.listeningDecorator(Executors.newSingleThreadExecutor());
+        ListenableFuture<Void> success = listeningExecutorService.submit(() -> null);
+        ListenableFuture<Void> failure = listeningExecutorService.submit(() -> {
+            throw new IllegalStateException();
+        });
```

Comment:
should use separate test cases for the successful/failed futures. Otherwise when tests fail, it. takes additional debugging to figure out what the problem is.